File: backend/app/api/v1/routes/handles.py
import uuid
from datetime import UTC, datetime, timedelta

from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.v1.deps import get_current_user
from app.core.database import get_db
from app.models.user import User
from app.schemas.handle import (
    HandleConfirmRequest,
    HandleInitiateRequest,
    HandleInitiateResponse,
    HandleResponse,
    HandleVerifiedResponse,
    SyncResponse,
)
from app.services.handle import (
    confirm_verification,
    get_handle_for_user,
    initiate_verification,
    list_handles,
    unlink_handle,
)
from app.workers.enqueue import enqueue_sync

router = APIRouter(prefix="/handles", tags=["handles"])


@router.get("", response_model=list[HandleResponse])
async def get_handles(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> list[HandleResponse]:
    handles = await list_handles(db, current_user.id)
    return [HandleResponse.model_validate(h) for h in handles]


@router.post("/verify/initiate", response_model=HandleInitiateResponse, status_code=status.HTTP_201_CREATED)
async def initiate(
    body: HandleInitiateRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> HandleInitiateResponse:
    row = await initiate_verification(db, current_user.id, body.handle, body.platform)
    return HandleInitiateResponse(
        handle_id=row.id,
        handle=row.handle,
        platform=row.platform,
        token=row.verification_token,
        expires_at=row.verification_token_expires_at,
    )


@router.post("/verify/confirm", response_model=HandleVerifiedResponse)
async def confirm(
    body: HandleConfirmRequest,
    background_tasks: BackgroundTasks,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> HandleVerifiedResponse:
    row = await confirm_verification(db, current_user.id, body.handle_id)
    enqueue_sync(row.id, background_tasks)
    return HandleVerifiedResponse(
        handle_id=row.id,
        handle=row.handle,
        platform=row.platform,
        verified_at=row.verified_at,
    )


@router.post("/{handle_id}/sync", response_model=SyncResponse)
async def manual_sync(
    handle_id: uuid.UUID,
    background_tasks: BackgroundTasks,
    force: bool = False,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> SyncResponse:
    handle = await get_handle_for_user(db, current_user.id, handle_id)

    # `force=true` lets the owner bypass the cooldown to recover a handle whose data
    # is missing/stale (e.g. a sync that stored 0 submissions). Owner-only via
    # get_handle_for_user, so it can't be abused against another account.
    cooldown = timedelta(minutes=30)
    if (
        not force
        and handle.last_manual_sync_at
        and datetime.now(UTC) - handle.last_manual_sync_at < cooldown
    ):
        retry_after = int((handle.last_manual_sync_at + cooldown - datetime.now(UTC)).total_seconds())
        raise HTTPException(
            status_code=status.HTTP_429_TOO_MANY_REQUESTS,
            detail={"message": "Sync cooldown active", "retry_after_seconds": retry_after},
        )

    handle.last_manual_sync_at = datetime.now(UTC)
    await db.commit()

    task_id = enqueue_sync(handle_id, background_tasks)
    return SyncResponse(task_id=task_id, handle_id=handle_id)


@router.delete("/{handle_id}", status_code=status.HTTP_204_NO_CONTENT)
async def unlink(
    handle_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> None:
    await unlink_handle(db, current_user.id, handle_id)

File: frontend/app/_lib/handles.ts
"use client";

import { apiFetch } from "./api";

export interface HandleData {
  id: string;
  handle: string;
  platform: string;
  is_verified: boolean;
  is_locked: boolean;
  status: string;
  sync_status: string;
  verified_at: string | null;
  last_synced_at: string | null;
  lockout_expires_at: string | null;
}

export interface InitiateData {
  handle_id: string;
  handle: string;
  platform: string;
  token: string;
  expires_at: string;
}

export interface VerifiedData {
  handle_id: string;
  handle: string;
  platform: string;
  verified_at: string;
}

export class ApiError extends Error {
  status: number;
  attemptsRemaining?: number;

  constructor(message: string, status: number, attemptsRemaining?: number) {
    super(message);
    this.status = status;
    this.attemptsRemaining = attemptsRemaining;
  }
}

export async function fetchHandles(token: string): Promise<HandleData[]> {
  const res = await apiFetch("/api/v1/handles", { token });
  if (!res.ok) throw new ApiError("Failed to fetch handles", res.status);
  return res.json();
}

export async function initiateVerification(
  token: string,
  handle: string,
): Promise<InitiateData> {
  const res = await apiFetch("/api/v1/handles/verify/initiate", {
    method: "POST",
    token,
    body: JSON.stringify({ handle, platform: "codeforces" }),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    const detail = err.detail ?? "Failed to initiate verification";
    throw new ApiError(
      typeof detail === "string" ? detail : detail.message ?? "Failed to initiate verification",
      res.status,
    );
  }
  return res.json();
}

export async function confirmVerification(
  token: string,
  handleId: string,
): Promise<VerifiedData> {
  const res = await apiFetch("/api/v1/handles/verify/confirm", {
    method: "POST",
    token,
    body: JSON.stringify({ handle_id: handleId }),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    const detail = err.detail ?? {};
    const message =
      typeof detail === "string"
        ? detail
        : detail.message ?? "Verification failed";
    const attemptsRemaining =
      typeof detail === "object" ? detail.attempts_remaining : undefined;
    throw new ApiError(message, res.status, attemptsRemaining);
  }
  return res.json();
}

export async function unlinkHandle(token: string, handleId: string): Promise<void> {
  const res = await apiFetch(`/api/v1/handles/${handleId}`, {
    method: "DELETE",
    token,
  });
  if (!res.ok && res.status !== 204) throw new ApiError("Failed to unlink handle", res.status);
}

export async function syncHandle(token: string, handleId: string): Promise<void> {
  const res = await apiFetch(`/api/v1/handles/${handleId}/sync`, {
    method: "POST",
    token,
  });
  // 429 = cooldown active, treat as ok (sync already running or recently ran)
  if (!res.ok && res.status !== 429) throw new ApiError("Failed to start sync", res.status);
}
